TGEN: PR for master branch since working state has been achieved

pemami4911 opened the pull request:
The TGEN is in a functional state; current support is for geometric RRTstar and RRTConnect. In the params launch file, you can select whichever planner you'd like to use. So far, without obstacle avoidance, RRTConnect runs faster and with higher quality paths. Next implementation steps involves the module that will take PCL data and form a traversability map that the TGEN can use for state validation. See the README for details concerning the new `sub8_msgs::PathPlan` service. 

See the commits from today (January 31st) for code review purposes.  The main work was removing unnecessary code related to system dynamics and updating the remaining code. 

Minor change was made to the C++ alarm system; the data structure containing AlarmRaisers was changed to a std::map to allow indexing by the individual alarm names.

pemami4911 committed:
TGEN: Added Eigen, finishing ODE

pemami4911 committed:
TGEN: Renamed files, refactored with Eigen

Adhering to naming convention where
sub8 is appending to file names that
contain extensions to OMPL classes

Finished refactoring with the Eigen
library

Added a TGenThrusterInfo struct that
encapsulates the process of
parsing the thruster info param
and setting the contents of
the thruster lever arms and thruster
directions matrices needed by the ODE
solver

Added the SubDynamics class
which maintains the thruster information
necessary for solving the sub ode

pemami4911 committed:
TGEN: Finished implementing ODE, added tests

Once we have the B matrix being
pushed out as a rosparam,
I will update the TGenThrusterInfo
constructor to grab the L and D
matrices. For now,
just leaving them empty and testing
with hardcoded values

pemami4911 committed:
TGEN: Grab thruster ranges and sets cspace bounds

The TGEN node now calls the thrusters/thruster_range
service during initialization and passes
along those values to the space_information_generator
to set the control space bounds. Updates unit
tests where appropriate.

pemami4911 committed:
TGEN: Adding post-propagate

pemami4911 committed:
TGEN: Updated distance function to use Eigen

pemami4911 committed:
TGEN: Warn before waiting for thrusters

pemami4911 committed:
TGEN: postPropagate, regularize rate change of orientation

Added a regularization term when computing the
rate change of orientation. If the current orientation
is not a unit quaternion, the added term will
attemptto drive it back to having a norm of 1.

Implemented the simple postPropagate method,
which re-normalizes the orientation. Both
this and the regularization should ensure
we keep our orientation-game on 100

jpanikulam committed:
Merge pull request #34 from pemami4911/tgen

TGEN: Adding new feature branch and latest commits

Approver: Jacob Panikulam

pemami4911 committed:
Merging local master into TGEN feature branch

pemami4911 committed:
TGEN: Added tgen alarms

Added alarms by utilizing the new
C++ alarm generation code

pemami4911 committed:
TGEN: Merge branch 'master' into tgen after setup.py hotfix

Conflicts:
	command/sub8_alarm/CMakeLists.txt
	command/sub8_alarm/include/sub8_alarm/alarm_helpers.h

pemami4911 committed:
ALARMS: Fix for failing unit tests on semaphore

pemami4911 committed:
TGEN: Integrating new C++ alarm system

Moving all configuration parameters
to the ros param server and
adding a "Planning failed" alarm

pemami4911 committed:
MISC: Updating package.xmls to format 2

Cleans up the build/run depend tags

Also fixed some errors with the sub8_alarm
CMakeLists.txt to properly export headers

pemami4911 committed:
TGEN: rosparam server and integration test changes

Cleaned up the CMakeLists

Now using rosparam server for all constants
that need to be set at runtime

Added a method that allows for the positional planning region
to be updated at will (In case the spatial planning region
will be changing on new updates to a traversability map)

Added integration test for SpaceInformationGenerator since
it relies on the rosparam server

Added alarm initialization code

pemami4911 committed:
PROPULSION: Adding B matrix service

"There is no spoon?"
"Then you'll see, that it is not the spoon that bends,
it is only yourself."

pemami4911 committed:
TGEN: Automatically pull and parse B matrix

B matrix service added to the thruster
mapper in order for the TGEN and other
nodes to have access to it. The B matrix
is passed as a flattened matrix in row-major
order.

Integration tests were added and corresponding
components of the TGEN were updated

pemami4911 committed:
TGEN: Addressing comments in PR #41

jpanikulam committed:
Merge pull request #41 from pemami4911/tgen

TGEN: New rostests, rosparams, B matrix service for thruster mapper

Approved: Jacob Panikulam

pemami4911 committed:
TGEN: Minor updates to prepare for integration tests

pemami4911 committed:
TGEN: Implemented a more sophisticated distance function

In preparation for end-to-end testing of the TGEN,
I updated the distance function to a more sophisticated
metric that uses a weighted euclidean distance between
two states. See the implementation for more details.

Also did some minor refactoring to the Sub8StateSpace.

Updated unit and integration tests to reflect
all changes.

pemami4911 committed:
MERGING: Merge branch 'master' into tgen

Conflicts:
	command/sub8_alarm/CMakeLists.txt
	gnc/sub8_trajectory_generator/CMakeLists.txt

pemami4911 committed:
TGEN: CMakeLists.txt fixes from merge errors

jpanikulam committed:
Merge pull request #47 from pemami4911/tgen

TGEN: Improved distance function, refactoring, and new tests

Approver: Jacob Panikulam

pemami4911 committed:
MERGING: Merge branch 'master' into tgen

Conflicts:
	gnc/sub8_thruster_mapper/nodes/mapper.py

pemami4911 committed:
MERGING: Merge branch 'master' into tgen

pemami4911 committed:
TGEN: Node name bug fix

Name of node was changed in
CMakeLists.txt but not elsewhere,
which was causing problems

pemami4911 committed:
TGEN: Removed thread lock from mapper.py

Removed the thread lock from mapper.py, which
was causing the mapper.py to get blocked when
we didn't want it to.

Cleaned up some other TGEN code.

pemami4911 committed:
TGEN: Refactoring to support geometric planning

The TGEN now supports geometric planning
instead of control-based planning. Essentially,
this means that the TGEN takes in only a start pose
and end pose, and calculates the position and orientation
(yaw) at each point along the path. See sub8_msgs::PathPlan.srv

pemami4911 committed:
ALARMS: Changed the AlarmRaiser data structure

Using a std::map instead of a std::vector so
the alarms can be called by name instead of trying
to remember the order they were stored in the vector (yikes!)

All alarm tests passing

pemami4911 committed:
Merge branch 'master' of https://github.com/uf-mil/Sub8 into tgen

pemami4911 committed:
Merge branch 'tgen' of https://github.com/uf-mil/Sub8 into tgen

jpanikulam commented on command/sub8_alarm/include/sub8_alarm/alarm_helpers.h:
```diff
@@ -3,13 +3,13 @@
-#ifndef ALARM_HELPERS_H_
-#define ALARM_HELPERS_H_
+#ifndef _ALARM_HELPERS_H
+#define _ALARM_HELPERS_H
```
No-Change: You can use #pragma once to replace header guards

jpanikulam commented on gnc/sub8_trajectory_generator/src/tgen_manager.cc:
```diff
@@ -0,0 +1,257 @@
+  ros::param::get("alarms_dir", alarms_dir);
+  std::string pkg_path = ros::package::getPath("sub8_trajectory_generator");
+  char file_sep = '/';
+#ifdef _WIN32
```
?

pemami4911 replied:
C++ alarms allow the user to specify alarms with JSON, by placing them in an "alarms" directory. I'm gonna move most of this code to the alarm helper class though in a new commit.

jpanikulam replied:
Sorry - poor expression; Why ifdef _WIN32?

jpanikulam commented on gnc/sub8_trajectory_generator/test/tgen_manager_test.cc:
```diff
@@ -0,0 +1,119 @@
+  ompl::base::StateSpacePtr _test_space;
+};
+
+TEST_F(TGenManagerTest, testSetProblemDefinition) {
```
What's this testing?

pemami4911 replied:
This is just a test fixture for creating an instance of the TGenManager class for each test

jpanikulam commented on gnc/sub8_trajectory_generator/test/tgen_integration.test:
```diff
@@ -0,0 +1,8 @@
+        args="--configuration_path=$(find sub8_videoray_m5_thruster)/config/calibration.json" />
+  <include file="$(find sub8_thruster_mapper)/launch/thruster_mapper.launch" />  
+  <include file="$(find sub8_trajectory_generator)/launch/trajectory_generator_params.launch" />
+  <test test-name="tgen_integration" name="tgen_integration_test" pkg="sub8_trajectory_generator" type="sub8_trajectory_generator_integration_tests" time-limit="60.0"/>
```
Do the integration tests verify that the tgen can take PathPlan service requests and yield a path?

pemami4911 replied:
Currently no, will add one in a new commit

jpanikulam replied:
Ok - will pull on that.

jpanikulam commented:
Misc comments for future:
- Try to squash sequential merges from master (db82812 and 5a50b44)
- 910a944 looks like a redundant merge?

Very pristine for a 2,000 line change commit!

pemami4911 committed:
TGEN: Updates for PR #54 - Alarm fixes and new test

Added tests to ensure planning happens successflly

Moved some alarm code for C++ alarms from the client side
to the lib

jpanikulam commented:
Ready to pull once merge conflicts are resolved.

pemami4911 committed:
MERGE: Merge branch 'tgen' with origin/master

pemami4911 committed:
MERGE: Merge branch 'master' of https://github.com/uf-mil/Sub8 into tgen

Conflicts:
	gnc/sub8_thruster_mapper/nodes/mapper.py
